src/NServiceBus.Core/Performance/Metrics/ReceiveDiagnosticsBehavior.cs

```diff
@@ -0,0 +1,63 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Metrics;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class ReceiveDiagnosticsBehavior : IBehavior<IIncomingPhysicalMessageContext, IIncomingPhysicalMessageContext>
+    {
+        static readonly Meter NServiceBusMeter = new Meter(
+            NServiceBusDiagnosticsInfo.InstrumentationName,
+            NServiceBusDiagnosticsInfo.InstrumentationVersion);
+
+        static readonly Counter<long> TotalProcessedSuccessfully =
+            NServiceBusMeter.CreateCounter<long>("messaging.successes", description: "Total number of messages processed successfully by the endpoint.");
+
+        static readonly Counter<long> TotalFetched =
+            NServiceBusMeter.CreateCounter<long>("messaging.fetches", description: "Total number of messages fetched from the queue by the endpoint.");
+
+        static readonly Counter<long> TotalFailures =
+            NServiceBusMeter.CreateCounter<long>("messaging.failures", description: "Total number of messages processed unsuccessfully by the endpoint.");
+
+        public ReceiveDiagnosticsBehavior(string endpointName, string queueNameBase, string discriminator)
+        {
+            this.endpointName = endpointName;
+            this.queueNameBase = queueNameBase;
+            this.discriminator = discriminator ?? "";
+        }
+
+        public async Task Invoke(IIncomingPhysicalMessageContext context, Func<IIncomingPhysicalMessageContext, Task> next)
+        {
+            context.MessageHeaders.TryGetMessageType(out var messageTypes);
```

Comment:
this is a bit of an indication that we're trying to gather information at a stage too early in the pipeline.
The alternative would be to set a "placeholder" class in the context and then in the logical stage add the necessary values so we can read them once we're back from calling `next` but for a single value it might be a bit too much